Fix source type binding key on windows

BoykoAlex opened the pull request:
On Windows since path separator is `\` the binding key ends up being `Lp/<path to the java file>~X`, i.e. includes file path. On Linux and Mac due to separator being `/` the key is `Lp/X;` as expected.
PR should fix the binding key on Windows

BoykoAlex commented:
Could someone please look into merging this PR? It's rather simple fix with a unit test... (cc: @martinlippert @jukzi)

jukzi commented on org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/dom/StandAloneASTParserTest.java:
```diff
@@ -1961,7 +1961,73 @@ class com {
+		packageDir.mkdir();
+		File file = new File(packageDir, "X.java");
+		Writer writer = null;
+		try {
```
please use try-with-resource

jukzi commented on org.eclipse.jdt.core.tests.compiler/src/org/eclipse/jdt/core/tests/dom/StandAloneASTParserTest.java:
```diff
@@ -1961,7 +1961,73 @@ class com {
+		}
+
+		try {
+			final String canonicalPath = file.getCanonicalPath();
```
why do you need super slow canonicalPath here? is absolute path enough?

jukzi replied:
this question was not answered

martinlippert commented:
Thanks a lot @jukzi for jumping on this and for providing feedback, very much appreciated. Look like @BoykoAlex incorporated your suggestions already. Do you think this is ready for merging into master?

iloveeclipse commented:
Merge commits are not allowed. Please rebase your changes on top of master branch instead.

BoykoAlex commented:
Squashed and rebased. Should be good to merge now.

jukzi commented:
> Look like @BoykoAlex incorporated your suggestions already. Do you think this is ready for merging into master?

does not look like. Or did it get lost during merge?

martinlippert commented:
> > Look like @BoykoAlex incorporated your suggestions already. Do you think this is ready for merging into master?
> 
> does not look like. Or did it get lost during merge?

Weird. I think I saw those questions addressed in the code, but maybe I am wrong here. I am sure @BoykoAlex can shed some light on this... :-)

BoykoAlex commented:
I don't know how it lost the additional "corrections" commit. I squashed 2 commits into one. I see that indeed those changes are gone. I've just added the changes again myself and pushed additional commit to my branch but the PR isn't showing this additional commit :-\

jukzi commented:
you need to force push

jukzi commented:
Or create a new PR, if you having trouble.
